### PR TITLE
feat: check if coin already in cache

### PR DESCRIPTION
**Pull request by hal3e**
# Summary
Checks if any of the inputs is already cached before sending/submitting the tx.

### CODE REVIEW

File: packages/fuels-accounts/src/coin_cache.rs [hunk 1]
```diff
@@ -41,6 +41,23 @@ impl CoinsCache {
...
+                    if items.contains(&CoinCacheItem::new(id.clone())) {
+                        return Some((key.clone(), id.clone()));
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
```
[the remark this answers is not in the document]
Comment (hal3e) [reply packages/fuels-accounts/src/coin_cache.rs:60]: sure

**Review by segfault-magnet — APPROVED** [review]

**Review by segfault-magnet — DISMISSED** [review]

**Review by digorithm — DISMISSED** [review]

File: packages/fuels-accounts/src/provider.rs [hunk 2]
```diff
@@ -233,9 +237,55 @@ impl Provider {
...
+    ) -> Result<()> {
+        use fuels_core::types::errors::{transaction, Error};
+
+        if let Some(((addr, asset_id), coin_type_id)) = self.find_in_cache(coin_ids).await {
+            let msg = match coin_type_id {
+                CoinTypeId::UtxoId(utxo_id) => format!("coin with utxo_id: `{utxo_id:x}`"),
+                CoinTypeId::Nonce(nonce) => format!("message with nonce: `{nonce}`"),
+            };
+            Err(Error::Transaction(transaction::Reason::Validation(
+                format!("{msg} already in cache. Wallet address: `{addr}`, asset id: `{asset_id}`"),
```
Comment (MujkicA) [inline packages/fuels-accounts/src/provider.rs:277]: I feel like just stating that the coin is already cached doesn't convey enough info to the user as to what is going on, especially if they are not fully aware of the internals of the SDK and the caveats of a UTXO based chain.

Saying something like:

"a transaction with this input was submitted recently, attempting to spend it again will result in an error".

should give a better idea of what the issue is.

**Review by MujkicA — COMMENTED** [review]
IIRC the plus here is that we prevent the user from hitting the network with a tx that's bound to fail, which is great.

I have two concerns, correct me if I'm wrong:
- We might get inconsistent results because of the time based eviction. E.g. user stalls submitting the second transaction long enough for the cache to drop the entry. 
- We don't evict when txs are dropped by the node which means we might prevent the user from executing a valid transaction at some point.

**Comment by hal3e** [comment]
> IIRC the plus here is that we prevent the user from hitting the network with a tx that's bound to fail, which is great.
> 
> I have two concerns, correct me if I'm wrong:
> 
>     * We might get inconsistent results because of the time based eviction. E.g. user stalls submitting the second transaction long enough for the cache to drop the entry.
> 
>     * We don't evict when txs are dropped by the node which means we might prevent the user from executing a valid transaction at some point.
- If the cache drops the entry because of time based eviction then we revert to the old behavior. We send the the tx and the node will complain.
- This can also happen with the current implementation; however, only if the tx is `squeezed out`.

* Had meeting with @segfault-magnet and @MujkicA where we went through the uses cases and we came to the above conclusions.

**Review by segfault-magnet — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]

**Review by digorithm — APPROVED** [review]

**Pull request merged**